### PR TITLE
Added on exit function to cleanup LVGL environment

### PR DESCRIPTION
**Pull request by vanwinkeljan**
Added function to LVGL glue logic to run cleanup on exit of the native posix board.

This provides the option to run a native posix application with valgrind to detect potential memory leaks.

### CODE REVIEW

**Comment by ioannisg** [comment]
Is this needed for 2.0, @vanwinkeljan ?

**Comment by vanwinkeljan** [comment]
@ioannisg no, 2.1.0 or latter is fine

File: lib/gui/lvgl/lvgl.c
Comment (aescolar) [inline lib/gui/lvgl/lvgl.c]: compiling samples/gui/lvgl/
```
lib/gui/lvgl/lvgl.c: In function ‘lvgl_cleanup’:
lib/gui/lvgl/lvgl.c:201:17: warning: unused variable ‘disp_buf’ [-Wunused-variable]
  lv_disp_buf_t *disp_buf;
                 ^~~~~~~~
```

**Review by aescolar — CHANGES_REQUESTED** [review]
Stopping it with Ctrl+C it dies miserably:
```
***** Booting Zephyr OS build v1.9.0-16805-gad2d065f0306 *****
[00:00:00.000,000] <inf> lvgl: lv_init ready
[00:00:00.000,000] <inf> lvgl: Screen create ready
[00:00:00.000,000] <inf> lvgl: Screen create ready
[00:00:00.000,000] <inf> lvgl: Screen create ready
[00:00:00.000,000] <inf> lvgl: label created
[00:00:00.000,000] <inf> lvgl: label created
^C
Stopped at 1.010s
programming error: posix_irq_handler_im_from_sw called from a HW model thread
programming error: posix_irq_handler_im_from_sw called from a HW model thread
programming error: posix_irq_handler_im_from_sw called from a HW model thread
programming error: posix_irq_handler_im_from_sw called from a HW model thread
programming error: posix_irq_handler_im_from_sw called from a HW model thread
programming error: posix_irq_handler_im_from_sw called from a HW model thread
programming error: posix_irq_handler_im_from_sw called from a HW model thread
ASAN:DEADLYSIGNAL
=================================================================
==11330==ERROR: AddressSanitizer: SEGV on unknown address 0xbbbbbbc7 (pc 0x565933f6 bp 0xffda70b8 sp 0xffda70a0 T0)
==11330==The signal is caused by a READ memory access.
    #0 0x565933f5 in lv_img_decoder_close /lvgl/src/lv_draw/lv_img_decoder.c:177

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV lvgl/src/lv_draw/lv_img_decoder.c:177 in lv_img_decoder_close
==11330==ABORTING
```

**Comment by vanwinkeljan** [comment]
> Stopping it with Ctrl+C it dies miserably

Will have a closer look has something to do with libc heap vs kernel space heap object allocation

**Comment by vanwinkeljan** [comment]
@aescolar looks like `k_free` calls `irq_unlock` which ends up in `posix_irq_handler_im_from_sw`.
`posix_irq_handler_im_from_sw` notices that it is no longer running in _sw_ context and raises an error followed by an exit call which ends up again in the `k_free`, end result is a crash.

Should `posix_exit` maybe be updated to abort on such a condition instead of recursively calling itself?

The crash itself has been fixed by only including the cleanup code if the libc heap is used.

**Comment by aescolar** [comment]
> looks like k_free calls[..]

You cannot use this mechanism in this way. At this point everything is being teared down and there is no Zephyr kernel anymore. So you cannot call any Zephyr code. See this note:

https://github.com/zephyrproject-rtos/zephyr/blob/204c9971ab535005858a069a6f70639eddba9f89/soc/posix/inf_clock/soc.h#L41-L42

This is a very native_posix only hook. You can only use this mechanism to close/clean up host OS resources that you allocated outside of Zephyr context.

**Comment by vanwinkeljan** [comment]
> You cannot use this mechanism in this way

Ok I do understand this but but my question was more if we should protect `posix_exit` so that it cant be called multiple times (btw will not do it in this PR, just wondering)?

> You can only use this mechanism to close/clean up host OS resources that you allocated outside of Zephyr context.

PR has been updated to only clean up the resources allocated by LVGL via the OS libc malloc and free, so that the memory is not reported as lost by Valgrind/ASAN.
In case Zephyrs memory management infrastructure is used it will not run any cleanup but this is not a problem as neither Valgrind or ASAN are aware of it.

**Comment by aescolar** [comment]
@vanwinkeljan 

> if we should protect posix_exit so that it cant be called multiple times (btw will not do it in this PR, just wondering)?

(AFAIR) Well.. `posix_exit()` is meant to be re-called. If in some way it gets called from a Zephyr thread (say an ASSERT or fault), it will actually block that pthread and awake the HW models pthread to call `posix_exit()` again.
When it gets called from a HW thread, it will pthread_cancel all Zephyr threads, and then (only from the HW thread) call the registered native tasks ON_EXIT. Meaning, the ON_EXIT tasks will always be called from the HW thread.

> In case Zephyrs memory management infrastructure is used it will not run any cleanup but this is not a problem as neither Valgrind or ASAN are aware of it.

Yep, the whole Zephyr "heap" is just a statically allocated chunk of RAM. So from valgrind's point of view it is just a normal static array with some random content the program chooses to put there.
If you have an actual heap leak in your Zephyr app you will just run out of Zephyr heap. valgrind can't help you there with its leak detection.

> by LVGL via the OS libc malloc

Why are you allocating from LVGL with the host malloc? (can you point to that code)

**Comment by vanwinkeljan** [comment]
> (AFAIR) Well.. posix_exit() is meant to be re-called. ...

Thx for all the info!

> Why are you allocating from LVGL with the host malloc? (can you point to that code)

LVGL can be configured to either use kernel heap, libc heap, a dedicated kernel space memory pool or a dedicated user space memory pool.

In case libc heap is selected `malloc` and `free` are used; on _real_ targets this calls are either handled by minimal libc or  newlibc, in case of native posix it is handled by the host libc (see commit: https://github.com/zephyrproject-rtos/zephyr/commit/95cd063280b848f9399272ca7a4c7db8c256977c)

**Comment by aescolar** [comment]
> LVGL can be configured to either use kernel heap, libc heap, a dedicated kernel space memory pool or a dedicated user space memory pool.

Ok, I understand.. I would discourage using the libC malloc in app code in native_posix, unless you really know what you are doing (e.g. trying to debug a leak in your code with valgrind).
And encourage developers to always have LVGL (or any other library or app) run as close as possible as to how it would run in an embedded target.
If you start to diverge in what a library or app does for native_posix you will start to see different bugs in each, which defeats the point of native_posix.

>In case libc heap is selected malloc and free are used; on real targets this calls are either handled by minimal libc or newlibc, in case of native posix it is handled by the host libc (see commit: 95cd063)

Sigh.. I know.. And I was just hoping not many people would use malloc but use k_malloc instead (I never got to add it to posix_cheats as I tend to recall it was not used in-tree at the time, but if you are really using it in LVGL maybe I should add it outside of the POSIX_API ifdef). This devolves into the whole POSIX API discussions on native_posix (which as of today is just disabled). malloc is not something one would normally expects to have from a embedded C library (not a baremetal one that is). It goes well into the OS domain, and into the POSIX API compliant OS part.

**Comment by vanwinkeljan** [comment]
> ... but if you are really using it in LVGL maybe I should add it  ...

I was just toying around so no need to put the effort in now.

On this PR, what do you think of the current proposal of only performing the cleanup incase libc is used for LVGL? This will not cause any direct issues as it is actually only cleaning up an external module through the modules interface and not using any specific zephyr calls.

**Comment by aescolar** [comment]
> and not using any specific zephyr calls.

Even if that was the case with the PR as it is now, my concern would be how to ensure that is the case in the future. This is calling into Zephyr application/library code. So it is breaking that separation, and will only work if you ensure that code being called remains being "safe" for this purpose. => After this PR would be merged you would need to police in some way that nobody modifies the LVGL code (or any code it uses) to do something which would break the assumption that nothing in Zephyr can be assumed to be running.

Note that in general the ON_EXIT cleanup code should assume it may have been triggered after a badly crashed Zephyr system (e.g. k_panic ), and that any Zephyr data may be corrupted.

Please tell me if I'm misunderstanding something, but I would not merge this PR, but just let that memory not be freed when using libC. (Anyhow the SDL library is leaving behind 4MB, and all of it is free'd by the Linux Kernel afterwards. The only issue here is the valgrind leak report being full of noise)

**Comment by vanwinkeljan** [comment]
> but I would not merge this PR

OK no problem then I will close this PR, no use in letting it dangle forever.

> The only issue here is the valgrind leak report being full of noise

This was the only reason, note that I swapped out tye SDL display by the _dummy_ display driver to get rid of the SDL stuff

**Pull request closed without merging**
